WIP: Refactor IAST::FormatSettings

murfel opened the pull request:
# Template
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
IAST::FormatSettings refactoring to encapsulate its fields. Now proper API added to be used mostly within IAST::formatImpl, primarily used for clickhouse-format.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

# My comments
 #45534

This PR will consist of several commits, each commit will encapsulate one field at a time. Each commit is supposed to keep the build stable.

## Accidentally fixed issues
1. `show_secrets` was missed in semi-copy-ctor - https://github.com/ClickHouse/ClickHouse/pull/45649/commits/c0bcf9808106dd08c8c22ca444ce420c2b4a1da0
2. ~10 hiliting issues that I've found while refactoring #45668

murfel committed:
Minor fix: add forgotten show_secrets to FormatSettings's ctor

murfel committed:
IAST::FormatSettings refactoring - hide always_quote_identifiers, identifier_quoting_style fields

murfel commented:
##  Issue # 3

The word 'identifier' in the second query is not hilited. Though it should be hilited either as an identifier or an alias. Probably an alias.

![image](https://user-images.githubusercontent.com/5155395/214866448-87823fc7-ffdc-4cc7-a8c9-4aec4ab909ce.png)

```
./clickhouse-format --hilite --query "with 2 as alias select * from table"
WITH 2 AS alias
SELECT *
FROM table

./clickhouse-format --hilite --query "with identifier as (select * from table) select * from table"
WITH identifier AS
    (
        SELECT *
        FROM table
    )
SELECT *
FROM table

./clickhouse-format --hilite --query "with identifier as alias select * from table"
WITH identifier AS alias
SELECT *
FROM table
```

murfel committed:
re-run CI

murfel committed:
IAST::FormatSettings refactoring - hide hilite

murfel committed:
Hide hilite amend
- hilite matching bracket
- remove accidentally added space
- remove unused field in Hiliter

murfel committed:
Hide one_line

murfel committed:
Hide show_secrets

vitlibar commented on src/Parsers/IAST.h:
```diff
@@ -181,29 +181,84 @@ class IAST : public std::enable_shared_from_this<IAST>, public TypePromotion<IAS
+            identifier_quoting_style(other.identifier_quoting_style), hilite(other.hilite),
+            should_show_secrets(other.should_show_secrets) {}
+
+        void writeKeyword(std::string_view str) const;
```
A thought about this PR a bit. `settings.writeKeyword()` just doesn't feel right. The `FormatSettings` was a simple data structure, and now it's going to become a complex class. The code seems to become more complex and difficult to understand than it was before.

I read how you came to this. But I suggest another approach - let's keep `FormatSettings` simple and move those function to a new class Formatter along with `ostr`.

```
class IAST
{
...
    class Formatter
    {
    public:
        Formatter(WriteBuffer & out, const FormatSettings & settings, FormatState & state);
        Formatter(const Formatter & src); // make Formatter copyable
        void setNeedParent(...)
        void writeFunction(std::string_view name) const;
        ...
    private:
        WriteBuffer & out; // settings.ostr moved here
        const FormatSettings & settings;
        FormatState & state;
        
        // Fields of FormatStateStacked:
        UInt8 indent = 0;
        bool need_parens = false;
        ...
    };
    
    void format(WriteBuffer & out, const FormatSettings & settings) const
    {
        FormatState state;
        formatImpl(Formatter{out, settings, state);
    }
    
    virtual void formatImpl(const Formatter & formatter) const;
};
```

Thus this new class will be instead of FormatStateStacked and it must be copyable - which means for example `ASTFunction::formatImpl` will be able to copy it, change something and call `formatImpl` on arguments with the changed formatter.

murfel replied:
This is basically taking the initial idea to the next level.

I'm all in for it, as it is quite confusing to have **three** formatting settings in `formatImpl`.

I'll take a look into the other two classes later and will be back.

murfel committed:
remove unneeded includes

murfel commented:
I have started implementing merging the FormatSettings, FormatState, and FormatStateStacked together, and now have a couple questions.

1. Design question. In `IAST.h` there are two functions, `format` and `formatImpl`. On the first glance, `format` creates FormatState and calls `formatImpl` and each implementation of `formatImpl` calls `formatImpl` on its children. 

```
void format(const FormatSettings & settings) const
{
    FormatState state;
    formatImpl(settings, state, FormatStateStacked());
}

virtual void formatImpl(const FormatSettings & /*settings*/, FormatState & /*state*/, FormatStateStacked /*frame*/) const
{
    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown element in AST: {}", getID());
}
```

However, there are about ~30 cases when `formatImpl` calls `format`, resetting the `FormatState`. E.g. 
https://github.com/ClickHouse/ClickHouse/blob/master/src/Parsers/Access/ASTSettingsProfileElement.cpp#L101

a) Why these usages need a blank state?

b) Since we're incapsulating FormatSettings, FormatState, and FormatStateStacked into a new Formatter class, both of `format` and `formatImpl` will have the same argument -- `Formatter`. It's now up to the user to pass a Formatter with a reset state. Thus we will only need one function, `format`. Please confirm if my understanding for the need of `format` and `formatImpl` is correct, and I could do that without causing something else to break?

2. C++ / api design question.

My API aspirations for the Formatter class is to delete all its copy/move c-tors/assignement operators and provide sereval methods for creating a copy of the formatter, e.g. `Formatter copy();` to copy with a blank state, `Formatter  copyWithIndent(int);` to copy with a specific indent set. All these copy methods will be based on the existing usage.

The benefit of this approach compared with just copy-contructors is so that you can have this:
```
void formatImpl(Formatter & out) {
...
    type->formatImpl(out.copyWithIndent(0));
...
}
```

Instead of this:
```
void formatImpl(Formatter & out) {
...
    Formatter stacked_formatter(out);
    stacked_formatter.setIndent(0);
    type->formatImpl(stacked_formatter);
...
}
```

(Adapted from https://github.com/ClickHouse/ClickHouse/blob/master/src/Parsers/ASTColumnDeclaration.cpp#L70)

However, since `formatImpl` takes Formatter by reference, it cannot bind to the temporary value of the copied Formatter in the first snipped.

Solution 1. Create a local variable to store a copy of `Formatter`, then pass it into 

Solution 2. Wrap it into a `unique_ptr` -- this will create an extra unnecesary code at each call.

Solution 3. I could just take it by r-value reference in `formatImpl`, but it would be morally wrong, as `Formatter` keeps references to `WriteBuffer`, `FormatSettings`, and `FormatState`, and this r-value object shouldn't be possibly able to live longer than the life of the stack that it was created on.

Solution 4. Make `Formatter` a **const** reference in `formatImpl`, this will require making all the internal fields references. Specifically, all the fields of `FormatStateStacked`. This is bad because it is an intentional confusion of the reader, as we do very much 'modify' Formatter through its references, and having references in classes in general is just such a mislead. However I would prefer this solution, as it creates a more consize code in the usage.

murfel committed:
Add FormattingBuffer
- Encapsulates FormatSettings, FormatState, FormatStateStacked
- Change IAST.format and IAST.formatImpl signatures
- 7/450 usages fixed (grep -rn "formatImpl") with dependencies
- Committed files pass static checks

murfel committed:
~50% formatImpl usages fixed

murfel committed:
~75% formatImpl usages fixed

murfel committed:
~90% formatImpl usages fixed

murfel committed:
~92% formatImpl usages fixed

murfel committed:
Fix formatImpl overrides calling format* w/o states
- Here all formatImpl overrides call a format* function within itself with just the FormatSettings argument, without the FormatState and FormatStateStacked. We need to make sure we pass a copy of the FormattingBuffer, to erase the state.
- TODO(me): check possible behaviour change, since we're keeping FormatState in the FormattingBuffer copy. I.e. check that those format*(FormatSettings) functions do not use or modify FormatState members.

murfel committed:
Fix terminal format* methods w/o state

murfel committed:
Fix non-terminal format* methods w/o state
- out.copy() (possibly?) needed when calling other format* methods w/o state

murfel committed:
Fix formatImpl* non-trivial calls in ASTFunction

murfel committed:
FormattingBuffer - add more copy* and setters/getters

murfel commented:
TODO:
- decide on the `IAST::format` signature
- `FormattingBuffer`
    - decide what `copy*()` should copy - `FormatSettings` always, `FormatState`? `FormatStateStacked`?
    - re-design `copy*()` and setters/getters
- pull master
- test

murfel commented:
TODO: write an explanation about squashing FormatSettings, FormatState, and FormatStateStacked.

murfel commented:
TODO: write down how I'm going to test this

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

# Conflicts:
#	src/Parsers/ASTColumnsTransformers.cpp
#	src/Parsers/ASTConstraintDeclaration.h
#	src/Parsers/ASTCreateQuery.cpp
#	src/Parsers/ASTCreateQuery.h
#	src/Parsers/ASTDeleteQuery.cpp
#	src/Parsers/ASTDictionary.h
#	src/Parsers/ASTFunction.cpp
#	src/Parsers/ASTIndexDeclaration.h
#	src/Parsers/ASTProjectionDeclaration.h
#	src/Parsers/ASTRenameQuery.h
#	src/Parsers/ASTSelectQuery.cpp
#	src/Parsers/ASTSetQuery.cpp
#	src/Parsers/ASTTableOverrides.h
#	src/Parsers/Access/ASTSetRoleQuery.h
#	src/Parsers/Access/ASTShowGrantsQuery.h
#	src/Parsers/IAST.h

murfel committed:
Merge branch 'master' into formatimpl-refactoring

murfel committed:
Merge branch 'ClickHouse:master' into formatimpl-refactoring

murfel commented:
> Why `stacked_state` is a unique ptr? You could just write

I can't keep `FormatStateStacked` as a value field because `formatImpl` accepts `FormattingBuffer` (my rename of Formatter) as a const-reference, thus I won't be able to modify any fields, but I could modify the pointed-to `FormatStateStacked`.

It's a bit ugly to have it as a unique_ptr, please see my comment above for the reasons we can't change the `formatImpl` signature to accept `FormattingBuffer` by value or by non-const reference.

> I understand why state is a shared pointer, but again it could be implemented without that:
Fair.

vitlibar commented:
> I can't keep FormatStateStacked as a value field because formatImpl accepts FormattingBuffer (my rename of Formatter) as a const-reference, thus I won't be able to modify any fields

Of course you can't modify a const object but you can modify a **copy** of a const object. See my example https://github.com/ClickHouse/ClickHouse/pull/45649#issuecomment-1468818520

murfel committed:
change IAST::format API and refactor (some, not all) usages

murfel committed:
undo making formatImpl protected for now

murfel committed:
fix one more format() usage

murfel committed:
make FormatSettings c-tor explicit

murfel commented:
> I understand why state is a shared pointer, but again it could be implemented without that:

Ah, I recalled. It's so that it doesn't need to be exposed to the user of `FormattingBuffer`. If it's a reference, a user has to pass it to the c-tor. Would you prefer to avoid a new call and have it as a reference?

murfel committed:
fix one more format() usage

murfel committed:
get rid of smart pointers in FormattingBuffer

murfel commented:
TODO: consider renaming auxiliary functions `formatOnCluster`, `formatSettingName` into `formatImpl*` for consistency.
(see src/Parsers/ASTAlterNamedCollectionQuery.cpp)

murfel commented:
TODO: files that require renaming format into formatImpl
src/Parsers/ASTBackupQuery.cpp

murfel commented:
Having written `child_node->formatImpl(out.copy())` a hundred times, I now feel like it's better to pass `FormattingBuffer` by value to `formatImpl` so that there's no need to call copy() every time.

The downsides are that there will be an unnecessary copy when we already prepared a copy to pass into formatImpl outside of its call, like in `ASTFunction.h`, search `out_need_parens`, `out_dont_need_parens` in my changes.

And I don't think there's any case at all, where we do want to keep the FormatStateStacked, rather than copy it for the child `formatImpl` call. The old API didn't allow it, so the new API doesn't need to allow it, either. Possibly, the new API should even prohibit it to prevent potential bugs.

I think we can spare a few extra copies, to avoid this `.copy()` repetition.

WDYT?

murfel commented:
Cataloguing format* usages which reset the state.

TODO: Decide, if it's on purpose. If not, use the already existing state.

FormatImpl 
https://github.com/murfel/ClickHouse/blob/master/src/Parsers/ASTFunction.cpp#L515

vitlibar commented:
> Having written child_node->formatImpl(out.copy()) a hundred times, I now feel like it's better to pass FormattingBuffer by value to formatImpl so that there's no need to call copy() every time.

Yes, passing by value should be also ok.

vitlibar commented:
> Should I also add an R-value overload as well?

Probably that won't work if there are two or more child nodes

murfel committed:
pass FormattingBuffer by value in formatImpl
formatImpl(const FormattingBuffer & -> formatImpl(FormattingBuffer

murfel committed:
pass FormattingBuffer by value in format*
format(.*)\(const FormattingBuffer & -> format$1\(const FormattingBuffer &

murfel committed:
pass FormattingBuffer by value in format*(..., const FB &)
format(.*)\((.*), const IAST::FormattingBuffer & -> format$1\($2, IAST::FormattingBuffer

murfel committed:
pass FormattingBuffer by value in formatImpl - two final cases
now const FormattingBuffer & doesn't exist in src/*

murfel committed:
revert deleting FB copy/move c-tor/op=

murfel committed:
revert copying FB when passing to format*
out.copy()) -> out)

murfel committed:
add .copy().setField() syntax & fix usages

murfel committed:
misc cleanup

murfel committed:
rename typo setNeedsParens -> setNeedParens

murfel committed:
replace internal format usages with formatImpl
(\.|(->))format\(out -> $1formatImpl(out ->

murfel committed:
implement copy/set methods in FB

murfel committed:
misc fixes in FB

murfel committed:
impl missed methods in FB

murfel committed:
fix format calls

murfel committed:
revert accidentally changed function name

murfel committed:
rename needsParens into needParens

murfel committed:
do not rename shouldShowSecrets

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

# Conflicts:
#	src/Parsers/Access/ASTGrantQuery.cpp

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

# Conflicts:
#	src/Parsers/Access/ASTGrantQuery.cpp

murfel committed:
fix missed usage in a merged file

murfel commented:
@vitlibar could you take a look again, please?

So I have left:

1. incapsulate FormattingBuffer.ostr (big change)
2. fix found hiliting bugs (https://github.com/ClickHouse/ClickHouse/issues/45668) in a separate PR
3. test

I'm going to work on 2 and 3 for now.

Do you think it would make sense to encapsulate FormattingBuffer.ostr in a separate PR? It doesn't seem to be tied to anything, and it's a 400-line change just to replace the usages of `out.ostr <<` into `out <<` and maybe 100-200 more to write changes in FormattingBuffer.

murfel committed:
fix my own bug: one_line setting was inverted

murfel committed:
fix my own bug: new line missed in expression list

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

murfel committed:
fix my own bug: extra_indent wasn't multiplied correctly

murfel committed:
fix my own bug: like argument output twice

murfel committed:
fix my own bug: like argument output twice, missed () in operators

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

murfel committed:
fix my own bug: missed using & with check

murfel committed:
fix my own bug: inverted nlOrWs and comment in create

murfel committed:
revert debug output in formatAST.cpp

murfel committed:
fix my own bug: only print nl in ASTExpressionList if not oneline

murfel committed:
fix my own bug: do not write db name as keyword in EXISTS DATABASE db_name

murfel committed:
fix my own bug: accidentally inverted condition in RENAME DICTIONARY

murfel committed:
fix my own bug (???): comma not hilited in ASTExpressionList

murfel committed:
add regression test script

murfel committed:
not a bug, revert hiliting comma in ASTExpressionList

vitlibar commented:
> Do you think it would make sense to encapsulate FormattingBuffer.ostr in a separate PR?

This PR is already big enough - if you can do that in a separate PR please do

vitlibar commented on tests/compare_hilites.sh:
```diff
@@ -0,0 +1,38 @@
+
+model_formatter="/home/natasha/MurfelClickHouse/build/programs/clickhouse-format";
+new_formatter="/home/natasha/copyClickHouse/build/programs/clickhouse-format";
+compare="/home/natasha/copyClickHouse/tests/a.out"
```
what is a.out ?

murfel replied:
That's the compiled hilite-aware comparator, it's in the same folder. I just committed this script as is for now because I'll probably need to change it anyway.

vitlibar replied:
`a.out` is meaningless, let's use `hilite-aware-comparator` instead

murfel committed:
fix my own bug: hilite comma when it's used as JoinKind::Comma

murfel committed:
fix my own wrong code formatting

vitlibar commented on tests/compare_hilites.sh:
```diff
@@ -0,0 +1,38 @@
+
+model_formatter="/home/natasha/MurfelClickHouse/build/programs/clickhouse-format";
+new_formatter="/home/natasha/copyClickHouse/build/programs/clickhouse-format";
+compare="/home/natasha/copyClickHouse/tests/a.out"
```
We usually use environmental variables or command line arguments to pass paths to a script which cannot be determined by that script. I mean `model_formatter`, `new_formatter`, `compare`, `flags`.

murfel committed:
remove friend FormattingBuffer from Hiliter

murfel committed:
copySettings -> getSettings

murfel committed:
delete FormatSettings c-tors in favor of aggregate initialization

murfel commented on src/Parsers/getInsertQuery.cpp:
```diff
@@ -19,10 +19,8 @@ std::string getInsertQuery(const std::string & db_name, const std::string & tabl
-    settings.always_quote_identifiers = true;
-    settings.identifier_quoting_style = quoting;
-    query.IAST::format(settings);
+    IAST::FormatSettings settings(true, true, quoting);
```
Fixed. A few notes:

Con: modern IDE should be able to inlay parameter names.
Pro: emacs [seems to have acquired](https://lists.gnu.org/archive/html/emacs-devel/2023-02/msg00841.html) such functionality just a month ago, and GitHub doesn't have it at all.

Con: now we can't have any c-tor for the struct (aggregate class shouldn't have any user-defined c-tors).
Pro: there's only one usage which needed a separate c-tor and it could be done differently.

Con: it's too verbose if initializing with variables: `.one_line = one_line`
Pro: inlay parameter names suggested by IDE aren't smart enough to avoid verboseness, too :(

All in all, the designated initializers win here :)

![image](https://user-images.githubusercontent.com/5155395/229646252-a91e5aed-3d1b-493c-992a-4580dd29bca5.png)

murfel committed:
Merge branch 'master' into formatimpl-refactoring

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

# Conflicts:
#	src/Parsers/ASTAlterQuery.cpp
#	src/Parsers/ASTSystemQuery.cpp

murfel commented:
@vitlibar please take a look again, everything related to the refactoring (rather than the script/comparator) is fixed.

murfel committed:
fix ASTUndropQuery

murfel committed:
fix ASTShowColumnsQuery

murfel committed:
avoid initializing FormatSettings with default values

murfel commented on tests/hilite_aware_comparator.cpp:
```diff
@@ -0,0 +1,109 @@
+
+const std::vector<const char *> HILITES =
+    {
+        "\033[1m", "\033[0;36m", "\033[0;33m", "\033[1;33m", "\033[0;32m", "\033[1;36m",
```
Fixed in #47610.

murfel commented on tests/hilite_aware_comparator.cpp:
```diff
@@ -0,0 +1,109 @@
+        CANCELLING_HILITE
+    };
+
+[[maybe_unused]] const char * consume_hilites(const char ** it)
```
Fixed in #47610.

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

# Conflicts:
#	src/Parsers/ASTCreateQuery.cpp
#	src/Parsers/ASTDictionary.cpp
#	src/Parsers/ASTDictionaryAttributeDeclaration.cpp
#	src/Parsers/ASTFunction.cpp
#	src/Parsers/ASTKillQueryQuery.cpp
#	src/Parsers/ASTWithAlias.cpp
#	src/Parsers/ASTWithElement.cpp

murfel committed:
delete old comparator & move script to its new place

murfel committed:
add CLI interface to comparator and cmake (not working yet)

vitlibar commented on src/Parsers/HiliteComparator/CMakeLists.txt:
```diff
@@ -5,3 +5,6 @@ add_headers_and_sources(hilite_comparator .)
 target_link_libraries(hilite_comparator PRIVATE dbms)
+
+clickhouse_add_executable(hilite_comparator_main hilite_comparator_main.cpp)
+target_link_libraries(hilite_comparator_main PRIVATE hilite_comparator)
```
"main" is not a clear name. I suggest renaming it into just "hilite_comparator" and the library into "hilite_comparator_library".

vitlibar commented on src/Parsers/HiliteComparator/compare_hilites.sh:
```diff
@@ -0,0 +1,38 @@
+
+model_formatter="/home/natasha/MurfelClickHouse/build/programs/clickhouse-format";
+new_formatter="/home/natasha/copyClickHouse/build/programs/clickhouse-format";
+compare="/home/natasha/copyClickHouse/tests/a.out"
```
Please change this line into

```
hilite_comparator="../../../build/src/Parsers/HiliteComparator/hilite_comparator"
```

and write a comment that in order to use this script it's necessary to first build this comparator along with `clickhouse-format` and that can be performed with commands like `cd build && ninja clickhouse-format hilite_comparator`.

murfel replied:
I've written hilite_comparator as an argument as well:
```
model_formatter=$1  # e.g. "ClickHouse/build/programs/clickhouse-format";
new_formatter=$2  # e.g. "refactorClickHouse/build/programs/clickhouse-format";
hilite_comparator=$3  # e.g. "refactorClickHouse/build/src/Parsers/HiliteComparator/hilite_comparator"
```

Instead of the relative path you've suggested, because we cannot assume where the bash script would be called from (as in 'which repo', and also as in which folder).

murfel committed:
fix my own duplication bug in ASTAlterQuery

murfel committed:
nit: make writeIdentifierOrAlias arg consistent with the name

murfel committed:
fix my own bug with an unneeded newline in "create table ... AS\n\n..."

murfel committed:
nit: consistent formatting for all isOneLine() usages

murfel committed:
move failing hilite test from tests/queries into gtest_format_hiliting.cpp

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

# Conflicts:
#	src/Parsers/IAST.h

vitlibar commented on src/Interpreters/Cache/QueryCache.cpp:
```diff
@@ -142,9 +142,7 @@ bool QueryCache::Key::operator==(const Key & other) const
-    IAST::FormatSettings format_settings(buf, /*one_line*/ true);
-    format_settings.show_secrets = false;
-    ast->format(format_settings);
+    ast->format(buf, IAST::FormatSettings{.show_secrets = false});
```
Can't we write just `ast->format(buf, {.show_secrets = false});` ? Because it's shorter

murfel replied:
Fixed.

However, it was a conscious decision to spell out `IAST::FormatSettings` so that it's more obvious to the reader, and the usages of `IAST::FormatSettings` are more searchable.

I do like your option better for readability. I guess, people can just look up `format` signature, and they could search for all `format` entries instead.

vitlibar commented on src/Parsers/ASTAlterNamedCollectionQuery.cpp:
```diff
@@ -12,41 +11,39 @@ ASTPtr ASTAlterNamedCollectionQuery::clone() const
-    settings.ostr << (settings.hilite ? hilite_identifier : "") << backQuoteIfNeed(collection_name) << (settings.hilite ? hilite_none : "");
-    formatOnCluster(settings);
+    out.writeKeyword("Alter NAMED COLLECTION ");
+    out.writeProbablyBackQuotedIdentifier(collection_name);
```
This is quite a long function name. Can it be shorter? And also "probably" doesn't sound very good (like "probably" or "probably not"?) "If needed" looks better

murfel replied:
The name is due to the function name which was used in place of all the usages of my newly introduced `writeProbablyBackQuotedIdentifier`.

https://github.com/murfel/ClickHouse/blob/69957e214762b884d759dc0169b87745367d1227/src/Parsers/IAST.cpp#L409

For consistency, I won't be fixing this.

But I was also confused by the numerous `write*` functions, and `probably` didn't help :(

vitlibar replied:
ok

vitlibar commented on src/Parsers/ASTAlterQuery.cpp:
```diff
@@ -127,364 +127,368 @@ const char * ASTAlterCommand::typeToString(ASTAlterCommand::Type type)
-                      << (settings.hilite ? hilite_none : "");
-        col_decl->formatImpl(settings, state, frame);
+        out.writeKeyword("ADD COLUMN ");
+        out.writeKeyword(if_not_exists ? "IF NOT EXISTS " : "");
```
We don't have to call `writeKeyword()` if `if_not_exists == false`:

```
if (if_not_exists)
    out.writeKeyword("IF NOT EXISTS ")
```

murfel replied:
This pattern of `out.writeKeyword(condition ? "a" : "b")` is used throughout the code, and I believe it's a nice and consistent way to output numerous optional modifiers. It's shorter in one line, and the odd `if` won't be breaking the flow of multiple writeKeywords visually.

vitlibar replied:
I don't see any sense in writing `hilite_keyword` and `hilite_none` with nothing in between.

vitlibar replied:
You could at least modify `writeKeyword()` and make it skip its argument if it's empty.

vitlibar commented on src/Parsers/ASTAlterQuery.cpp:
```diff
@@ -127,364 +127,368 @@ const char * ASTAlterCommand::typeToString(ASTAlterCommand::Type type)
-        column->formatImpl(settings, state, frame);
+        out.writeKeyword(clear_column ? "CLEAR " : "DROP ");
+        out.writeKeyword("COLUMN ");
+        out.writeKeyword(if_exists ? "IF EXISTS " : "");
```
We don't have to call writeKeyword() if if_not_exists == false

murfel replied:
Same, won't fix

vitlibar commented on src/Parsers/ASTAlterQuery.cpp:
```diff
@@ -127,364 +127,368 @@ const char * ASTAlterCommand::typeToString(ASTAlterCommand::Type type)
-        settings.ostr << " " << (settings.hilite ? hilite_none : "");
-        comment->formatImpl(settings, state, frame);
+        out.writeKeyword("COMMENT COLUMN ");
+        out.writeKeyword(if_exists ? "IF EXISTS " : "");
```
and here

murfel replied:
Same, won't fix

vitlibar commented on src/Parsers/ASTAlterQuery.cpp:
```diff
@@ -569,58 +573,51 @@ ASTPtr ASTAlterQuery::clone() const
-        frame_nested.expression_list_always_start_on_new_line = true;
-        command_list->as<ASTExpressionList &>().formatImplMultiline(settings, state, frame_nested);
-    }
+    FormattingBuffer copy = out.copy().setNeedParens(false);
```
It seems you don't need to copy `out` here because earlier in this function you already set `NeedParens` to false and also because this function owns `out` and can change it.

vitlibar commented on src/Parsers/ASTCheckQuery.h:
```diff
@@ -26,29 +26,28 @@ struct ASTCheckQuery : public ASTQueryWithTableAndOutput
             }
-            settings.ostr << (settings.hilite ? hilite_keyword : "") << indent_str << backQuoteIfNeed(getTable()) << (settings.hilite ? hilite_none : "");
+            out.writeIndent();
+            out.writeKeyword(backQuoteIfNeed(getTable()));
```
It seems a bit strange that database and table are formatted as keywords, but I see in master it was the same.

murfel replied:
Added this to the collection of formatting / hiliting issues #48140

vitlibar commented on src/Parsers/ASTColumnsTransformers.cpp:
```diff
@@ -315,23 +313,24 @@ void ASTColumnsReplaceTransformer::Replacement::updateTreeHashImpl(SipHash & has
 {
-    settings.ostr << (settings.hilite ? hilite_keyword : "") << "REPLACE" << (is_strict ? " STRICT " : " ") << (settings.hilite ? hilite_none : "");
+    out.writeKeyword("REPLACE");
+    out.writeKeyword(is_strict ? " STRICT " : " ");
```
The same here

murfel replied:
Won't fix

murfel committed:
fix HiliteComparator compilation

vitlibar commented on src/Parsers/ASTCreateQuery.cpp:
```diff
@@ -329,127 +334,133 @@ void ASTCreateQuery::formatQueryImpl(const FormatSettings & settings, FormatStat
+            columns_list->formatImpl(out);
+            out.nlOrNothing();
+            out.ostr << ")";
+            out.setExpressionListAlwaysStartsOnNewLine(false);
```
If you copy `out` at the beginning of the scope then you won't need to restore the original value of `ExpressionListAlwaysStartsOnNewLine` at the end of the scope.

vitlibar commented on src/Parsers/ASTFunction.cpp:
```diff
@@ -522,31 +522,28 @@ void ASTFunction::appendColumnNameImpl(WriteBuffer & ostr) const
-            FormatStateStacked frame;
             writeCString("(", ostr);
-            window_definition->formatImpl(format_settings, state, frame);
+            window_definition->format(ostr, FormatSettings());
```
Just `window_definition->format(ostr, {});` is shorter

vitlibar commented on src/Parsers/ASTFunction.cpp:
```diff
@@ -622,51 +619,17 @@ ASTPtr ASTFunction::toLiteral() const
-  *
-  * Another case is regexp match. Suppose the user types match(URL, 'www.clickhouse.com'). It often means that the user is unaware that . is a metacharacter.
-  */
-static bool highlightStringLiteralWithMetacharacters(const ASTPtr & node, const IAST::FormatSettings & settings, const char * metacharacters)
```
Why did `highlightStringLiteralWithMetacharacters` disappear?

murfel replied:
Moved to `FormattingBuffer` because it's called when `hilite` flag is set, but I've encapsulated the hilite flag, and don't want to expose a getter for just this one usage.

vitlibar commented on src/Parsers/ASTInsertQuery.cpp:
```diff
@@ -110,30 +105,41 @@ void ASTInsertQuery::formatImpl(const FormatSettings & settings, FormatState & s
+        if (settings_ast)
+            out.nlOrWs();
+        else
+            out.ostr << ' ';
```
Now it became more complicated. We can add a lambda function `put_delimiter()` and write here

```
put_delimiter();
select->formatImpl(out);
```

murfel committed:
fix hilite_comparator_main compilation

murfel committed:
compile comparator against clickhouse_parsers instead of dbms

murfel committed:
rename hilite_comparator -> _library, _main -> hilite_comparator

murfel committed:
script: add a docstring, add arguments, rename the compare variable

murfel committed:
get rid of excessive IAST::FormatSettings

murfel commented on src/Parsers/ASTAlterQuery.cpp:
```diff
@@ -127,364 +127,368 @@ const char * ASTAlterCommand::typeToString(ASTAlterCommand::Type type)
-                      << (settings.hilite ? hilite_none : "");
-        col_decl->formatImpl(settings, state, frame);
+        out.writeKeyword("MODIFY COLUMN ");
+        out.writeKeyword(if_exists ? "IF EXISTS " : "");
```
Same, won't fix

murfel committed:
avoid copy in ASTAlterQuery.cpp

murfel committed:
more FormatSettings() -> {}

murfel committed:
add put_delimiter in ASTInsertQuery.cpp

murfel committed:
Merge remote-tracking branch 'origin/master' into formatimpl-refactoring

vitlibar commented:
Too much hiliting. For example, let's consider the query `SHOW TABLES`. Now with your changes it's displayed as

```
\u001b[1m (hilite_keyword)
SHOW
\u001b[0m (hilite_none)

\u001b[1m (hilite_keyword)
\u001b[0m (hilite_none)

\u001b[1m (hilite_keyword)
TABLES
\u001b[0m (hilite_none)
```

And before your changes it was just

```
\u001b[1m (hilite_keyword)
SHOW TABLES
\u001b[0m (hilite_none)
```

The test [02105_backslash_letter_commands](https://s3.amazonaws.com/clickhouse-test-reports/45649/14bf688c5d3be378815add1704c0110336586c92/fast_test.html) failed because of that.

vitlibar commented on src/Parsers/IAST.cpp:
```diff
@@ -210,14 +200,166 @@ String IAST::getColumnNameWithoutAlias() const
+void IAST::FormattingBuffer::writePossiblyHilited(std::string_view str, const char * hilite_type) const
+{
+    Hiliter hiliter = createHiliter(hilite_type);
+    ostr << str;
```
A possible optimization: instead of writing a closing hiliter in `~Hiliter` `FormattingBuffer` can remember the last opening hiliter and check it with next output. For example, if you have a sequence of `writeKeyword()` it would allow to write only one `hilite_keyword` and only one `hilite_none`.

murfel replied:
I was thinking about but introducing a state machine would complicate the logic for no gain, as there's no need for the optimisation of hilite. There are so many public methods which will need to flush the buffer.

And if anyone ever goes back to implementing the original task of limiting line length, then it will become even more complicated, as it will probably need a state machine of its own.

vitlibar replied:
> And if anyone ever goes back to implementing the original task of limiting line length, then it will become even more complicated, as it will probably need a state machine of its own.

What is the original task?

murfel replied:
[Limit line length in clickhouse-format command #38317
](https://github.com/ClickHouse/ClickHouse/issues/38317)

vitlibar replied:
In regard to https://github.com/ClickHouse/ClickHouse/issues/38317 I still don't see why it's better to format
```
\u001b[1mSHOW \u001b[0m\u001b[1m\u001b[0m\u001b[1mTABLES\u001b[0m
```
instead of
```
\u001b[1mSHOW TABLES\u001b[0m
```

In order to limit the length of a line the characters must be counted without hilite sequences anyway, I suppose. So `FormattingBuffer` needs to memorize the current number of characters in the current line. And I don't think there is a lot of difference here between memorizing one or two variables.

murfel replied:
I feel like it's a premature optimisation. It doesn't affect performance, it only complicates the code logic.

I agree that squashing hilites together does improve readability for debugging, though. But it's a very abstract scenario.

However, I just realised that implementing squashing might over-complicate the whole PR. Earlier we agreed that I'll encapsulate `FormattingBuffer.ostr` in a separate PR because it affects ~1000 lines.

To summarize, here are pros for squashing hilites:
- easier readability for debugging
- XX% fewer bytes per hilited query (but does anyone care?)

Cons:
- complicates the code (additional state machine)
- complicates the PR (encapsulating `ostr` here)
- I'll need to write more code instead of keeping everything as is :)

The question is, do you feel like the pros outweigh the cons? I'm honestly not very excited to write even more code at this point, and I feel like just merging this PR as is would already bring value.

vitlibar replied:
> complicates the PR (encapsulating ostr here)

We can encapsulate ostr in a half-way manner:

```
class FormattingBuffer
{
public:
    ~FormattingBuffer() { setHilite(hilite_none); }

    WriteBuffer & out() { setHilite(hilite_none); return ostr; }

    void writeKeyword(std::string_view str) const
    {
        if (keyword.empty())
            return;
        setHilite(hilite_keyword);
        ostr << str;
    }

private:
    WriteBuffer & ostr;
    ...
};
```

and also we will need to mechanically replace every `.ostr` with `.out()`. It doesn't seem like a difficult task.

> I feel like it's a premature optimisation. It doesn't affect performance, it only complicates the code logic.

I'm not 100% sure about performance, but anyway. We format ASTs very often, it's not like one-time task. So it would be nice to be accurate with that.

vitlibar commented on tests/queries/0_stateless/01316_create_user_syntax_hilite.reference:
```diff
@@ -1 +0,0 @@
-[1mCREATE USER[0m user[1m IDENTIFIED[0m[1m WITH plaintext_password[0m[1m BY[0m 'hello'
```
Let's restore this test. Later we can think about removing it but currently it's better to see the changes in hiliting this PR makes.

murfel replied:
I've moved this test here

https://github.com/murfel/ClickHouse/blob/2c1a311478b202724a84abd5a80a94f4b66709c8/src/Parsers/tests/gtest_format_hiliting.cpp#L127

where it will be tested with a hilite-aware comparator

vitlibar replied:
It's not exactly the same, so I think it's better to keep both tests for now.

vitlibar commented on src/Parsers/IAST.h:
```diff
@@ -240,13 +222,117 @@ class IAST : public std::enable_shared_from_this<IAST>, public TypePromotion<IAS
+        void nlOrWs() const;  // Write newline or whitespace.
+        void nlOrNothing() const;
+
+        void writeSecret(const String & secret = "") const;
```
Let's use `std::string_view` for such arguments always in this class where it doesn't cause troubles.

vitlibar commented:
> For example, let's consider the query `SHOW TABLES`. Now with your changes it's displayed as
> 
> ```
> \u001b[1m (hilite_keyword)
> SHOW
> \u001b[0m (hilite_none)
> 
> \u001b[1m (hilite_keyword)
> \u001b[0m (hilite_none)
> 
> \u001b[1m (hilite_keyword)
> TABLES
> \u001b[0m (hilite_none)
> ```
> 
> And before your changes it was just
> 
> ```
> \u001b[1m (hilite_keyword)
> SHOW TABLES
> \u001b[0m (hilite_none)
> ```

Here I just tried to make my comment more readable, the test's output was like

```
\u001b[1mSHOW \u001b[0m\u001b[1m\u001b[0m\u001b[1mTABLES\u001b[0m
```

alexey-milovidov commented:
@vitlibar This is a good pull request, and we need it.

vitlibar commented:
> @vitlibar This is a good pull request, and we need it.

Ok, I'll do rebasing of the PR and maybe also correct it a little bit.

murfel commented:
> > @vitlibar This is a good pull request, and we need it.
> 
> Ok, I'll do rebasing of the PR and maybe also correct it a little bit.

Seems like I won't be working on this after all, @vitlibar feel free to grab this.

vitlibar commented:
> Dear @murfel, this PR hasn't been updated for a while. Will you continue working on it? If not, please close it. Otherwise, ignore this message.

I'll continue working on this PR.

alexey-milovidov commented:
We need this!

vitlibar commented:
I'll split this PR into parts and merge it one by one:
[part 1](https://github.com/ClickHouse/ClickHouse/pull/45649)

rschu1ze commented:
Done in #72723 and #73163.
